data_structures/dsu_union_rank.cpp

```diff
@@ -0,0 +1,148 @@
+        vector<int> depth; ///<tracks the depth(rank) of i in the tree
+        vector<int> setSize;///<size of each chunk(set)
+    public:
+        /**
+         * @brief constructor for initialising all data members 
+         * @param n number of elements
+         */
+        explicit dsu(int n){
+            p.assign(n,0);
+            //initially all of them their own parents
+            depth.assign(n,0);
+            setSize.assign(n,0);
+            for(int i=0;i<n;i++){
+                p[i] = i;
+                depth[i] = 0;
+                setSize[i] = 1;
+            }
+        }
+        /**
+         * @brief Method to find the representative of the set to which i belongs to, T(n) = O(logN)
+         * @param i element of some set
+         * @returns representative of the set to which i belongs to
+         */
+        int findSet(int i){
+            /// using union-rank
+            while(i!=p[i]){
+                i = p[i];
+            }
+            return i;
+        }
+        /**
+         * @brief Method that combines two disjoint sets to which i and j belongs to and make 
+         * a single set having a common representative.
+         * @param i element of some set
+         * @param j element of some set
+         * @returns void
+         */
+        void unionSet(int i,int j){
+            //check if both belongs to same set or not
+            if(isSame(i,j)){
+                return;
+            }
+            //we find representative of the i and j
+            int x = findSet(i);
+            int y = findSet(j);
+
+            //always keeping the min as x
+            //in order to create a shallow tree
+            if(depth[x]>depth[y]){
+                std::swap(x,y);
+            }
+            //making the shallower tree' root parent of the deeper root
+            p[x] = y;
+
+            //if same depth then increase one's depth
+            if(depth[x] == depth[y]){
+                depth[y]++;
+            }
+            //total size of the resultant set
+            setSize[y]+=setSize[x];
+        }
+        /**
+         * @brief A utility function which check whether i and j belongs to same set or not
+         * @param i element of some set
+         * @param j element of some set
+         * @returns `true` if element i and j are in same set
+         * @returns `false` if element i and j are not in same set
+         */
+        bool isSame(int i,int j){
+            if(findSet(i) == findSet(j)){
+                return true;
+            }
+            return false;
+        }
+        /**
+         * @brief Method to print all the parents of i, or the path from i to representative.
+         * @param i element of some set
+         * @returns void
+         */
+        void getParents(int i){
+            while(p[i]!=i){
+                cout << i << " ->";
+                i = p[i];
+            }
+            cout << p[i] << endl;
+        }
+        
+};
+
```

Comment:
Add self-test implementations using `assert` (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program as an example).

Reply:
Done!